lrauv_ignition_plugins/src/DopplerVelocityLog.cc

```diff
@@ -0,0 +1,1213 @@
+            this->dataPtr->beamLobesMessage.mutable_marker(2 * i);
+        lobeConeMarkerMessage->set_parent(
+            this->dataPtr->depthSensor->Parent()->Name());
+        gz::msgs::Set(lobeConeMarkerMessage->mutable_scale(), scale);
+        gz::msgs::Set(lobeConeMarkerMessage->mutable_pose(),
+                      this->dataPtr->depthSensor->LocalPose() *
+                      beam.Transform());
+        auto * lobeCapMarkerMessage =
+            this->dataPtr->beamLobesMessage.mutable_marker(2 * i + 1);
+        lobeCapMarkerMessage->set_parent(
+            this->dataPtr->depthSensor->Parent()->Name());
+        gz::msgs::Set(lobeCapMarkerMessage->mutable_scale(), scale);
+        gz::msgs::Set(lobeCapMarkerMessage->mutable_pose(),
+                      this->dataPtr->depthSensor->LocalPose() *
+                      beam.Transform());
+      }
+
+      if (this->dataPtr->visualizeBeamReflections)
+      {
+        gz::math::Pose3d orientation;
+        orientation.Rot().SetFrom2Axes(
+            beam.Axis(), beamTarget->pose.Pos());
+        const auto scale = beamRange * gz::math::Vector3d::One;
+        auto * refConeMarkerMessage =
+            this->dataPtr->beamReflectionsMessage.mutable_marker(2 * i);
+        refConeMarkerMessage->set_parent(
+            this->dataPtr->depthSensor->Parent()->Name());
+        gz::msgs::Set(refConeMarkerMessage->mutable_scale(), scale);
+        gz::msgs::Set(
+            refConeMarkerMessage->mutable_pose(),
+            this->dataPtr->depthSensor->LocalPose() *
+            orientation * beam.Transform());
+        auto * refCapMarkerMessage =
+            this->dataPtr->beamReflectionsMessage.mutable_marker(2 * i + 1);
+        refCapMarkerMessage->set_parent(
+            this->dataPtr->depthSensor->Parent()->Name());
+        gz::msgs::Set(refCapMarkerMessage->mutable_scale(), scale);
+        gz::msgs::Set(
+            refCapMarkerMessage->mutable_pose(),
+            this->dataPtr->depthSensor->LocalPose() *
+            orientation * beam.Transform());
+      }
+
+      // Use shortest beam range as target range
+      targetRange = std::min(targetRange, beamRange);
+
+      // Compute target speed as it would have been measured by the beam
+      const EntityKinematicState & targetEntityStateInWorldFrame =
+          this->dataPtr->worldState[beamTarget->entity];
+
+      const gz::math::Pose3d targetPoseInWorldFrame =
+          sensorStateInWorldFrame.pose *
+          this->dataPtr->beamsFrameTransform *
+          beamTarget->pose;
+
+      const gz::math::Vector3d targetVelocityInWorldFrame =
+          targetEntityStateInWorldFrame.linearVelocity +
+          targetEntityStateInWorldFrame.angularVelocity.Cross(
+              targetPoseInWorldFrame.Pos() -
+              targetEntityStateInWorldFrame.pose.Pos());
+
+      const gz::math::Vector3d relativeSensorVelocityInSensorFrame =
+          sensorStateInWorldFrame.pose.Rot().RotateVectorReverse(
```

Comment:
@arjo129 I'll start by saying that this plugin implements bottom-tracking and reports ship-referenced velocity data (check the BS command in [RDI Pathfinder's manual](https://www.manualslib.com/manual/1447288/Teledyne-Pathfinder.html?page=233#manual)). That is, it measures the sensor linear velocity w.r.t. the bottom (or whatever is right under the sensor) in a frame that is aligned with that of the sensor itself i.e. the typical x-forward, y-left, z-up frame (well, not quite, you'll find a reference frame is specified to match [RDI Pathfinder conventions](https://www.manualslib.com/manual/1447288/Teledyne-Pathfinder.html?page=41#manual)). 

> IIUC we don't need to account for angular velocity of the vehicle because the linear velocity of the sensor should account for it?

That's correct. In particular, if the sensor frame were to be translated w.r.t. the vehicle frame, sensor linear velocity will account for the effects of vehicle angular velocity.

> Should we also account for aangular velocity of the sensor if for instance the axis of rotation of the vehicle is around the sensor itself?

I think not. Since our sensor has no volume, rotations about an axis that goes through the origin of the sensor frame do not translate the sensor (and so linear velocity remains to be zero). 

Having said that, projecting linear velocities onto beam lobes' symmetry axes is a very crude approximation to beam velocity estimates. I can imagine how changes in (a rotated projection of the) topography could look like a change in velocity to a beam. I have no idea what real DVLs do to deal with this.